torch/_dynamo/trace_rules.py

```diff
@@ -35,40 +44,121 @@
+    "torch.onnx.is_in_onnx_export": TorchInGraphFunctionVariable,
+    "torch.onnx.operators.shape_as_tensor": TorchInGraphFunctionVariable,
+    "torch.overrides.is_tensor_like": TorchInGraphFunctionVariable,
+    "torch.jit.is_scripting": TorchInGraphFunctionVariable,
+    "torch.jit.is_tracing": TorchInGraphFunctionVariable,
+    "torch.jit.annotate": TorchInGraphFunctionVariable,
+    "torch.distributed.is_available": TorchInGraphFunctionVariable,
+    "torch.distributed.is_initialized": TorchInGraphFunctionVariable,
+    "torch.distributed.get_rank": TorchInGraphFunctionVariable,
+    "torch.distributed.get_world_size": TorchInGraphFunctionVariable,
+    "torch.distributed._tensor.DTensor#from_local": TorchInGraphFunctionVariable,
+    "torch._utils.is_compiling": TorchInGraphFunctionVariable,
+    "torch.overrides.get_default_nowrap_functions": TorchInGraphFunctionVariable,
+    "torch.fx._symbolic_trace.is_fx_tracing": TorchInGraphFunctionVariable,
+    "torch._dynamo.external_utils.is_compiling": TorchInGraphFunctionVariable,
+    "torch.autograd.graph.disable_saved_tensors_hooks": TorchInGraphFunctionVariable,
+}
+
+
+auto_torch_name_rule_map = {
+    # Dynamo implemented context managers
     "torch._C.DisableTorchFunctionSubclass": TorchCtxManagerClassVariable,
     "torch.amp.autocast_mode.autocast": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.enable_grad": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.inference_mode": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.no_grad": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.set_grad_enabled": TorchCtxManagerClassVariable,
-    "torch.autograd.profiler.profile": TorchCtxManagerClassVariable,
-    "torch.autograd.profiler.record_function": TorchCtxManagerClassVariable,
     "torch.cpu.amp.autocast_mode.autocast": TorchCtxManagerClassVariable,
     "torch.cuda.amp.autocast_mode.autocast": TorchCtxManagerClassVariable,
-    "torch.profiler.profiler.profile": TorchCtxManagerClassVariable,
 }
 
 
+torch_name_rule_map = {**manual_torch_name_rule_map, **auto_torch_name_rule_map}
+
+
 @functools.lru_cache(None)
 def get_torch_obj_rule_map():
     d = dict()
     for k, v in torch_name_rule_map.items():
-        obj = load_object(k)
-        assert obj not in d
-        d[obj] = v
+        try:
+            obj = load_object(k)
+            d[obj] = v
+        except (AttributeError, ModuleNotFoundError):
+            pass
     return d
 
 
+def _load_obj_from_str(fully_qualified_name):
+    module, obj_name = fully_qualified_name.rsplit(".", maxsplit=1)
+    return getattr(importlib.import_module(module), obj_name)
+
+
 def load_object(name):
-    mod_name, obj_name = name.rsplit(".", 1)
-    mod = importlib.import_module(mod_name)
-    obj = getattr(mod, obj_name)
-    return obj
+    x = name.split("#")
+    if len(x) == 2:
+        obj = _load_obj_from_str(x[0])
+        val = getattr(obj, x[1])
+    else:
+        assert len(x) == 1
```

Comment:
put name in the assert msg for ease of debugging

Reply:
Yes